File: C/tim_sort.c
#include <stdio.h>
#include <stdlib.h>

#define RUN 32 

void insertionSort(int arr[], int left, int right) {
    for (int i = left + 1; i <= right; i++) {
        int temp = arr[i], j;
        for (j = i; j > left && arr[j - 1] > temp; j--)
            arr[j] = arr[j - 1];
        arr[j] = temp;
    }
}

void merge(int arr[], int l, int m, int r) {
    int len1 = m - l + 1, len2 = r - m;
    int left[len1], right[len2];

    for (int i = 0; i < len1; i++)
        left[i] = arr[l + i];
    for (int i = 0; i < len2; i++)
        right[i] = arr[m + 1 + i];

    int i = 0, j = 0, k = l;
    while (i < len1 && j < len2)
        arr[k++] = (left[i] <= right[j]) ? left[i++] : right[j++];

    while (i < len1)
        arr[k++] = left[i++];
    while (j < len2)
        arr[k++] = right[j++];
}

void timSort(int arr[], int n) {
    for (int i = 0; i < n; i += RUN)
        insertionSort(arr, i, (i + RUN - 1 < n - 1) ? i + RUN - 1 : n - 1);

    for (int size = RUN; size < n; size *= 2) {
        for (int left = 0; left < n; left += 2 * size) {
            int mid = left + size - 1, right = (left + 2 * size - 1 < n - 1) ? left + 2 * size - 1 : n - 1;
            if (mid < right)
                merge(arr, left, mid, right);
        }
    }
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int arr[] = {5, 21, 7, 23, 19};
    int n = sizeof(arr) / sizeof(arr[0]);

    timSort(arr, n);
    printf("Sorted array:\n");
    printArray(arr, n);

    return 0;
}

File: Kotlin/timSort.kt
fun minRunLength(n: Int): Int {
    var r = 0
    var length = n
    while (length >= 64) {
        r = r or (length and 1)
        length = length shr 1
    }
    return length + r
}

fun insertionSort(arr: IntArray, left: Int, right: Int) {
    for (i in left + 1..right) {
        val temp = arr[i]
        var j = i - 1
        while (j >= left && arr[j] > temp) {
            arr[j + 1] = arr[j]
            j--
        }
        arr[j + 1] = temp
    }
}

fun merge(arr: IntArray, left: Int, mid: Int, right: Int) {
    val len1 = mid - left + 1
    val len2 = right - mid
    val leftArr = IntArray(len1)
    val rightArr = IntArray(len2)

    for (i in 0 until len1) leftArr[i] = arr[left + i]
    for (i in 0 until len2) rightArr[i] = arr[mid + 1 + i]

    var i = 0
    var j = 0
    var k = left

    while (i < len1 && j < len2) {
        if (leftArr[i] <= rightArr[j]) {
            arr[k] = leftArr[i]
            i++
        } else {
            arr[k] = rightArr[j]
            j++
        }
        k++
    }

    while (i < len1) {
        arr[k] = leftArr[i]
        i++
        k++
    }

    while (j < len2) {
        arr[k] = rightArr[j]
        j++
        k++
    }
}

fun timSort(arr: IntArray) {
    val n = arr.size
    val minRun = minRunLength(n)

    for (i in 0 until n step minRun) {
        insertionSort(arr, i, kotlin.math.min(i + minRun - 1, n - 1))
    }

    var size = minRun
    while (size < n) {
        for (left in 0 until n step 2 * size) {
            val mid = left + size - 1
            val right = kotlin.math.min(left + 2 * size - 1, n - 1)

            if (mid < right) {
                merge(arr, left, mid, right)
            }
        }
        size *= 2
    }
}

fun main() {
    val arr = intArrayOf(5, 21, 7, 23, -7, 19, 0, 10, 3, 12, 8, -5, 15)
    timSort(arr)
    println("Sorted Array: ${arr.joinToString(", ")}")
}